### PR TITLE
feature importance type in saved model file

### PR DESCRIPTION
**Pull request by guolinke**

### CODE REVIEW

**Comment by StrikerRUS** [comment]
@guolinke Don't you want to expose this parameter to C API? With 3.0 I guess it's good time to break some code (if any). Then wrappers will be able to provide this param directly in function like start/end iterations 

https://github.com/microsoft/LightGBM/blob/2792923d35e7f3d392474ceffe2185c661047b2e/python-package/lightgbm/basic.py#L2734-L2750

**Comment by guolinke** [comment]
Thanks @StrikerRUS , I change all related functions.

File: R-package/R/lgb.Booster.R [hunk 1]
```diff
@@ -424,7 +424,7 @@ Booster <- R6::R6Class(
     },
 
     # Save model
-    save_model = function(filename, num_iteration = NULL) {
+    save_model = function(filename, num_iteration = NULL, feature_importance_type = 0) {
```
Comment (jameslamb) [inline R-package/R/lgb.Booster.R:427]: the R linter we use requires explicitly differentiating between integers and doubles.
```suggestion
    save_model = function(filename, num_iteration = NULL, feature_importance_type = 0L) {
```

File: R-package/R/lgb.Booster.R [hunk 2]
```diff
@@ -445,7 +446,7 @@ Booster <- R6::R6Class(
     },
 
     # Save model to string
-    save_model_to_string = function(num_iteration = NULL) {
+    save_model_to_string = function(num_iteration = NULL, feature_importance_type = 0) {
```
Comment (jameslamb) [inline R-package/R/lgb.Booster.R:449]:
```suggestion
    save_model_to_string = function(num_iteration = NULL, feature_importance_type = 0L) {
```

File: R-package/R/lgb.Booster.R [hunk 3]
```diff
@@ -457,12 +458,13 @@ Booster <- R6::R6Class(
...
         , private$handle
         , as.integer(num_iteration)
+        , as.integer(feature_importance_type)
       ))
 
     },
 
     # Dump model in memory
-    dump_model = function(num_iteration = NULL) {
+    dump_model = function(num_iteration = NULL, feature_importance_type = 0) {
```
Comment (jameslamb) [inline R-package/R/lgb.Booster.R:467]:
```suggestion
    dump_model = function(num_iteration = NULL, feature_importance_type = 0L) {
```

File: src/c_api.cpp [hunk 4]
```diff
@@ -689,21 +689,25 @@ class Booster {
...
     size_t len = std::strlen(model_str);
     boosting_->LoadModelFromString(model_str, len);
   }
 
-  std::string SaveModelToString(int start_iteration, int num_iteration) {
-    return boosting_->SaveModelToString(start_iteration, num_iteration);
+  std::string SaveModelToString(int start_iteration, int num_iteration,
+                                int feature_importance_type) {
+    return boosting_->SaveModelToString(start_iteration,
+                                        feature_importance_type, num_iteration);
```
Comment (guolinke) [inline src/c_api.cpp:704]:
```suggestion
                                        num_iteration, feature_importance_type);
```

**Comment by guolinke** [comment]
ping @imatiach-msft for swig fix

**Comment by imatiach-msft** [comment]
@guolinke this looks great!

**Comment by StrikerRUS** [comment]
@guolinke In my commit a2755330ab4ddf72254bd0e02ccc0b9eccfc478c I
replaced `cout` type with `split` type according to the existing terminology:
https://github.com/microsoft/LightGBM/blob/61b3c30842810205bc95736e6e66baa790b098cd/include/LightGBM/c_api.h#L1067-L1076
added defines in C API for better user experience and changed in Python package param type from int to string for better usage and to unify with existing codebase:
https://github.com/microsoft/LightGBM/blob/61b3c30842810205bc95736e6e66baa790b098cd/python-package/lightgbm/sklearn.py#L244-L247
https://github.com/microsoft/LightGBM/blob/61b3c30842810205bc95736e6e66baa790b098cd/python-package/lightgbm/basic.py#L2956-L2959

File: include/LightGBM/config.h [hunk 5]
```diff
@@ -532,6 +532,11 @@ struct Config {
   // desc = **Note**: can be used only in CLI version
   std::string output_model = "LightGBM_model.txt";
 
+  // desc = the feature importance type in the saved model file
+  // desc = ``0``: count-based feature importance (numbers of splits are counted); ``1``: gain-based feature importance (values of gain are counted)
+  // desc = **Note**: can be used only in CLI version
+  int saved_feature_importance_type = 0;
```
Comment (StrikerRUS) [inline include/LightGBM/config.h:538]: @guolinke I added `CLI-only` flag. Should it be marked as `[no-save]` as well?

**Comment by StrikerRUS** [comment]
Fixed #3011.

File: include/LightGBM/config.h (hunk 5 above)
Comment (guolinke) [reply include/LightGBM/config.h:538]: yeah, `CLI-only` is needed. As for `[no-save]`, I think it is better not marked, as user may want to know which feature importance type is saved in the model file.

**Comment by guolinke** [comment]
Thank you @StrikerRUS  for the doc improvement, as always!

File: include/LightGBM/config.h (hunk 5 above)
Comment (StrikerRUS) [reply include/LightGBM/config.h:538]: OK. Agree with you!

**Comment by StrikerRUS** [comment]
@jameslamb Is something left for R-package?

**Comment by StrikerRUS** [comment]
@jameslamb Thank you! I think it's better to have a separate PR for R tests.

**Pull request merged**